IPython/core/interactiveshell.py

```diff
@@ -1537,10 +1537,33 @@ def _ofind(self, oname, namespaces=None):
         Has special code to detect magic functions.
         """
         oname = oname.strip()
-        if not oname.startswith(ESC_MAGIC) and \
-                not oname.startswith(ESC_MAGIC2) and \
-                not all(a.isidentifier() for a in oname.split(".")):
-            return {'found': False}
+        raw_parts = oname.split(".")
+        parts = []
+        parts_ok = True
+        for p in raw_parts:
+            if p.endswith("]"):
+                var, *indices = p.split("[")
+                if not var.isidentifier():
+                    parts_ok = False
+                    break
+                parts.append(var)
+                for ind in indices:
+                    if ind[-1] != "]" and not ind[:-1].isnumeric():
```

Comment:
This currently prohibits `my_list[-1]`, `my_list[10_000]` or `my_list[0b101]`. One solution would be to use tokenize. I can provide a snippet. Other parts should not require a change for the first two cases since `int('-1')`, `int('10_000')` should just work. The binary/hex specification seems less important but could be allowed with a simple change too.

Reply:
No, it actually works ok, this code is much more permissive than I understood from reading it. It will actually accept anything.

Reply:
Yeah, it will allow it but will rarely receive it, the regex that transform `?` into `pinfo` is already restricted to 0-9. You can still call %pinfo directly.

Reply:
Could the regex also allow a single `-` prefix then? I often check the last item in the array and being able to check `mylist[-1]?` would be useful.

Reply:
Sure.